src/algebra/gradient.jl

```diff
@@ -251,12 +251,27 @@ function LazyOperators.materialize(
     lOp::Gradient{O, <:Tuple{LazyMapOver}},
     cPoint::CellPoint,
 ) where {O}
-    return MapOver(
-        materialize(
-            Base.Fix2(materialize, cPoint) ∘ Base.Fix2(Gradient, gradient_style(lOp)),
-            get_args(lOp)...,
-        ),
+    ∇x = tuplemap(
+        x -> materialize(Gradient(x, gradient_style(lOp)), cPoint),
+        get_args(get_args(lOp)...),
     )
+    return MapOver(∇x)
+end
+
+# Specialize the previous method when `Gradient` is applied
+# to two levels of LazyMapOver in order to help compiler inference
+# to deal with recursion.
+# This may be used by `bilinear_assemble` when a bilinear form
+# containing a `Gradient` is evaluated at a `CellPoint`
+function LazyOperators.materialize(
+    lOp::Gradient{O, <:Tuple{LazyMapOver{<:NTuple{N, LazyMapOver}}}},
+    cPoint::CellPoint,
+) where {O, N}
```

Comment:
Is it possible to put the content of this function (and the previous one) in a separate function (like `_materialize_gradient_cell_point`) ? It's weird to have the same code twice. If it's necessary for the full type inference then OK.

Reply:
I have already tried to factorize in a separate function but I haven't found how to do it without breaking performance (inlining doesn't seem to work because of the recursive calls )